core/src/main/scala/io/projectglow/transformers/normalizevariants/VariantNormalizer.scala

```diff
@@ -67,92 +68,372 @@ private[projectglow] object VariantNormalizer extends GlowLogging {
-        Option(ReferenceDataSource.of(Paths.get(refGenomePathString.get)))
-      } else {
-        None
-      }
+        val refGenomeDataSource = Option(ReferenceDataSource.of(Paths.get(refGenomePathString.get)))
 
-      it.flatMap { row =>
-        internalRowToVariantContextConverter.convert(row) match {
-
-          case Some(vc) =>
-            val vcStreamAfterMaybeSplit: Stream[VariantContext] =
-              // Strangely, Array, List, Seq, or Iterator do not work correctly!
-              // Line-by-line debugging shows that the issue is after the splitting.
-              // Assume a multi-allelic vc is split into two bi-allelic vcs. If the resulting two
-              // vcs are stored in any collection other than Stream, then the map function that
-              // applies the variantContextToInternalRowConverter.convertRow on them puts two
-              // copies of the InternalRow generated from the first vc in the result instead of one
-              // copy of each!
-              if (splitToBiallelic) {
-                splitVC(vc)
-              } else {
-                Stream(vc)
-              }
-
-            val isFromSplit = vcStreamAfterMaybeSplit.length > 1
-
-            val vcStreamAfterMaybeNormalize = if (doNormalize) {
-              vcStreamAfterMaybeSplit.map(vc =>
-                VariantNormalizer.normalizeVC(vc, refGenomeDataSource.get))
-            } else {
-              vcStreamAfterMaybeSplit
-            }
-
-            vcStreamAfterMaybeNormalize.map(
-              variantContextToInternalRowConverter.convertRow(_, isFromSplit)
-            )
-
-          case None => Stream(row)
+        it.map { row =>
+          val isFromSplit = row.getBoolean(splitFromMultiallelicColumnIdx)
+          internalRowToVariantContextConverter.convert(row) match {
 
+            case Some(vc) =>
+              variantContextToInternalRowConverter
+                .convertRow(VariantNormalizer.normalizeVC(vc, refGenomeDataSource.get), isFromSplit)
+            case None => row
+          }
         }
       }
+
+      SQLUtils.internalCreateDataFrame(df.sparkSession, rddAfterNormalize, schema, false)
+
+    } else {
+      dfAfterMaybeSplit
     }
 
-    SQLUtils.internalCreateDataFrame(df.sparkSession, mapped, schema, false)
+    dfAfterMaybeNormalize
 
   }
 
   /**
-   * Splits a vc to bi-alleleic using gatk splitting function
+   * Generates a new DataFrame by splitting the variants in the input DataFrame similar to what vt decompose does
    *
-   * @param vc
-   * @return a Stream of split vc's
+   * @param variantDf
+   * @return dataframe of split variants
    */
-  private def splitVC(vc: VariantContext): Stream[VariantContext] = {
-    GATKVariantContextUtils
-      .splitVariantContextToBiallelics(
-        vc,
-        false,
-        GenotypeAssignmentMethod.BEST_MATCH_TO_ORIGINAL,
-        false
+  @VisibleForTesting
+  private[normalizevariants] def splitVariants(variantDf: DataFrame): DataFrame = {
```

Comment:
Could we move the splitter into a separate file? It's pretty separate, and this file is getting long.

Reply:
Done